### PR TITLE
Homework 2

### PR DESCRIPTION
**Pull request by a-s01**

### CODE REVIEW

File: src/main/java/com/epam/spring/homework2/bean/BeanA.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package com.epam.spring.homework2.bean;
+
+import com.epam.spring.homework2.validation.annotation.IsInRange;
+import com.epam.spring.homework2.validation.annotation.IsNotNull;
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.factory.DisposableBean;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.beans.factory.annotation.Value;
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/BeanA.java]: unused import

File: src/main/java/com/epam/spring/homework2/bean/BeanF.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+package com.epam.spring.homework2.bean;
+
+import com.epam.spring.homework2.validation.annotation.IsInRange;
+import com.epam.spring.homework2.validation.annotation.IsNotNull;
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.factory.annotation.Autowired;
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/BeanF.java]: unused import

File: src/main/java/com/epam/spring/homework2/bean/ValidationRequiredAnnotationBeanPostProcessor.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
+package com.epam.spring.homework2.bean;
+
+import com.epam.spring.homework2.validation.ValidationException;
+import com.epam.spring.homework2.validation.Validator;
+import com.epam.spring.homework2.validation.ValidatorFactory;
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.BeansException;
+import org.springframework.beans.factory.BeanInitializationException;
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/ValidationRequiredAnnotationBeanPostProcessor.java]: unused import

File: src/main/java/com/epam/spring/homework2/bean/BeanB.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import com.epam.spring.homework2.validation.annotation.IsInRange;
+import com.epam.spring.homework2.validation.annotation.IsNotNull;
+import com.epam.spring.homework2.validation.annotation.ValidationRequired;
+import org.springframework.beans.factory.annotation.Value;
+
+@ValidationRequired
+public class BeanB {
+
+    @IsNotNull
+    @Value("${beanB.name}")
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/BeanB.java]: This is not a good approach. Please rework:
- make class immutable
- inject values from properties file in a method inside configuration class

Apply to all relevant classes

File: src/main/java/com/epam/spring/homework2/Main.java [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+public class Main {
+
+    public static void main(String[] args) {
+        AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(Config.class);
+
+        System.out.println("\nAll beans present in application context: ");
+        for (String beanDefinitionName : context.getBeanDefinitionNames()) {
+            System.out.println("- " + context.getBean(beanDefinitionName));
+        }
+
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/Main.java:15]: Missing item No 12

```java
System.out.println("\nAll beans definitions present in application context: ");
for (String beanDefinitionName : context.getBeanDefinitionNames()) {
    System.out.println("- " + context.getBeanDefinition(beanDefinitionName));
}
```

**Review by KostyantynPanchenko — CHANGES_REQUESTED** [review]
Please make fixes according to the comments.

Also configure `gitignore` to exclude `*.iml` and binaries. Examples are available in internet.

**Comment by a-s01** [comment]
Thanks for the review. 
I have fixed my issues. Please, check.

File: src/main/java/com/epam/spring/homework2/bean/BeanC.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public void initC() {
+        System.out.println("beanC: initC()");
+    }
+
+    public void destroyC() {
+        System.out.println("beanC: destroyC()");
+    }
+
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/BeanC.java]: extra line between the methods

File: src/main/java/com/epam/spring/homework2/bean/BeanD.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public void initD() {
+        System.out.println("beanD: initD()");
+    }
+
+    public void destroyD() {
+        System.out.println("beanD: destroyD()");
+    }
+
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/BeanD.java]: extra line between the methods

File: .gitignore [hunk 8]
```diff
@@ -1,2 +1,23 @@
...
+*.zip
+*.tar.gz
+*.rar
+
+### Java ###
+# Compiled class file
+*.class
+
+# Log file
+*.log
```
Comment (KostyantynPanchenko) [inline .gitignore]: Do you also see that red sign over here? :) It requires you to add a new line at the end of the file :)

File: src/main/resources/application.properties [hunk 9]
```diff
@@ -0,0 +1,6 @@
+beanB.name = B
+beanB.value = 1
+beanC.name = C
+beanC.value = 2
+beanD.name = D
+beanD.value = 3
```
Comment (KostyantynPanchenko) [inline src/main/resources/application.properties]: And here...

**Review by KostyantynPanchenko — CHANGES_REQUESTED** [review]
Please make changes according to the comments and have a great weekend

File: .gitignore (hunk 8 above)
Comment (a-s01) [reply .gitignore]: This just sucks. Sorry for it. I enabled the conforming setting in Intelijii, it should help me get rid of these forever. At least, I hope so :)

File: src/main/java/com/epam/spring/homework2/config/Config.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    @Value("${beanC.name}")
+    private String beanCName;
+    @Value("${beanC.value}")
+    private int beanCValue;
+    @Value("${beanD.name}")
+    private String beanDName;
+    @Value("${beanD.value}")
+    private int beanDValue;
+
+    @Bean(initMethod = "initD", destroyMethod = "destroyD")
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/config/Config.java]: Let's make the world beautiful with
```java
@Bean(initMethod = "initD", destroyMethod = "destroyD")
    public BeanD beanD(@Value("${beanD.name}") final String name, @Value("${beanD.value}") final int value) {
        return new BeanD(name, value);
    }
```
And get rid of those instance variables.

If we go even further - we can introduce a base class with getters, `toString()` and both `init` and `destroy` methods. Then we can make `BeanB`, `BeanC` and `BeanD` extend our base class ;)

**Comment by a-s01** [comment]
@KostyantynPanchenko would you mind checking this also? My bad, I didn't comment on my done work.

**Comment by KostyantynPanchenko** [comment]
Hi @a-s01 , changes are required for  `Config,java`

**Comment by a-s01** [comment]
@KostyantynPanchenko  I have double-checked and don't see any comment/change request on Config.java. Point it out to me, please. I have moved BeanB, BeanC and BeanD configuration to Config.class, but that was one commit ago...

**Comment by KostyantynPanchenko** [comment]
I am talking about this comment:
Screenshot 2022-02-07 225327

**Review by KostyantynPanchenko — APPROVED** [review]

**Comment by a-s01** [comment]
It's strange I didn't see the requested changes. Could it be because it was pending?
Anyway, thanks for the review!
Screenshot 2022-02-08 at 09 40 57

File: src/main/java/com/epam/spring/homework2/bean/BaseBean.java [hunk 11]
```diff
@@ -0,0 +1,31 @@
+package com.epam.spring.homework2.bean;
+
+import com.epam.spring.homework2.validation.annotation.IsInRange;
+import com.epam.spring.homework2.validation.annotation.IsNotNull;
+
+public abstract class BaseBean {
```
Comment (KostyantynPanchenko) [inline src/main/java/com/epam/spring/homework2/bean/BaseBean.java]: Does not contain abstract methods, thus should not be abstract. Also most likely can be package-private

**Review by KostyantynPanchenko — COMMENTED** [review]
See my comment in BaseBean

**Pull request merged**